### ISSUE TITLE
Regression from top-of-tree (2021-04-30)

### ISSUE THREAD

**Issue by eradman** [zenithdb/zenith#85]
I realize this is a new work-in-progress, but perhaps this would serve as a useful regression test

```sh
#!/bin/sh

export PGHOST=127.0.0.1
export PGPORT=55432
export PGDATABASE=postgres

[ -f chembl_28_postgresql.tar.gz ] || {
wget ftp://ftp.ebi.ac.uk/pub/databases/chembl/ChEMBLdb/latest/chembl_28_postgresql.tar.gz
tar -zxf chembl_28_postgresql.tar.gz
}
createuser chembl
createdb chembl_28 -O chembl
pg_restore -O -j 4 -U chembl -d chembl_28 chembl_28/chembl_28_postgresql/chembl_28_postgresql.dmp
```

While restoring I observed the following errors after trying to branching the database while the restore was running

```
% RUST_BACKTRACE=full ./target/debug/zenith branch eric
thread 'main' panicked at 'Missing start-point', zenith/src/main.rs:237:13
stack backtrace:
   0:        0x1030d008c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h9955000502fddd8a
   1:        0x1031068d0 - core::fmt::write::hd6aaffe828fad50f
   2:        0x1030cf7e4 - std::io::Write::write_fmt::h8939db9f062fb6a9
   3:        0x1030e663c - std::panicking::default_hook::{{closure}}::h22d061fa0f32cbb9
   4:        0x1030e61b0 - std::panicking::default_hook::h900d59958999a29b
   5:        0x1030e6af0 - std::panicking::rust_panic_with_hook::h407c595b7d5fe810
   6:        0x102c5a568 - std::panicking::begin_panic::{{closure}}::h160eaef961b35fb8
                               at /private/tmp/rust-20210325-61749-fsgy2a/rustc-1.51.0-src/library/std/src/panicking.rs:520:9
   7:        0x102c58bc0 - std::sys_common::backtrace::__rust_end_short_backtrace::hcd38a220f638de80
                               at /private/tmp/rust-20210325-61749-fsgy2a/rustc-1.51.0-src/library/std/src/sys_common/backtrace.rs:141:18
   8:        0x103108c70 - std::panicking::begin_panic::hd8fa8c5428944368
                               at /private/tmp/rust-20210325-61749-fsgy2a/rustc-1.51.0-src/library/std/src/panicking.rs:519:12
   9:        0x1027c2040 - zenith::run_branch_cmd::h257a5f0220d7d10f
                               at /Users/eradman/git.oss/zenith/zenith/src/main.rs:237:13
  10:        0x1027c02ec - zenith::main::h49c8481c8bce9d2a
                               at /Users/eradman/git.oss/zenith/zenith/src/main.rs:98:39
  11:        0x1027bccc0 - core::ops::function::FnOnce::call_once::hcf21086145e5d2a3
                               at /private/tmp/rust-20210325-61749-fsgy2a/rustc-1.51.0-src/library/core/src/ops/function.rs:227:5
  12:        0x1027c42cc - std::sys_common::backtrace::__rust_begin_short_backtrace::hb1bc45af75409a3c
                               at /private/tmp/rust-20210325-61749-fsgy2a/rustc-1.51.0-src/library/std/src/sys_common/backtrace.rs:125:18
  13:        0x1027bea10 - std::rt::lang_start::{{closure}}::hecd713fbcf38ff77
                               at /private/tmp/rust-20210325-61749-fsgy2a/rustc-1.51.0-src/library/std/src/rt.rs:66:18
  14:        0x1030ea1f0 - std::rt::lang_start_internal::hcd175898c32ebdc5
  15:        0x1027be9e8 - std::rt::lang_start::hc1dd096f18cf7119
                               at /private/tmp/rust-20210325-61749-fsgy2a/rustc-1.51.0-src/library/std/src/rt.rs:65:5
  16:        0x1027c3674 - _main
```

from `.zenith/pgdatadirs/pg1/log`:

```2021-04-30 10:46:18.442 EDT [23331] ERROR:  index "pg_toast_16481_index" contains unexpected zero page at block 0
2021-04-30 10:46:18.442 EDT [23331] HINT:  Please REINDEX it.
2021-04-30 10:46:18.442 EDT [23331] CONTEXT:  COPY component_sequences, line 92
2021-04-30 10:46:18.442 EDT [23331] STATEMENT:  COPY component_sequences (component_id, component_type, accession, sequence, sequence_md5sum, description, tax_id, organism, db_source, db_version) FROM stdin;
	
2021-04-30 10:46:18.788 EDT [23447] LOG:  [ZENITH_SMGR] libpqpagestore: connected to 'host=127.0.0.1 port=64000'
2021-04-30 10:46:18.843 EDT [23447] ERROR:  index "pk_actsm_id" contains unexpected zero page at block 0
2021-04-30 10:46:18.843 EDT [23447] HINT:  Please REINDEX it.
2021-04-30 10:46:18.843 EDT [23447] CONTEXT:  while cleaning up index "pk_actsm_id" of relation "public.activity_supp_map"
	automatic vacuum of table "chembl_28.public.activity_supp_map"
2021-04-30 10:46:18.855 EDT [23447] ERROR:  catalog is missing 1 attribute(s) for relid 58831
2021-04-30 10:46:18.855 EDT [23447] CONTEXT:  automatic vacuum of table "chembl_28.public.activity_supp"
TRAP: FailedAssertion("relid == targetRelId", File: "/Users/eradman/git.oss/zenith/tmp_install/build/../../vendor/postgres/src/backend/utils/cache/relcache.c", Line: 1090, PID: 23447)
0   postgres                            0x0000000104ef5de8 ExceptionalCondition + 268
1   postgres                            0x0000000104edee5c RelationBuildDesc + 324
2   postgres                            0x0000000104ede0f8 RelationIdGetRelation + 404
3   postgres                            0x00000001047858a8 try_relation_open + 192
4   postgres                            0x0000000104a5f714 vacuum_open_relation + 180
5   postgres                            0x0000000104a5e980 vacuum_rel + 372
6   postgres                            0x0000000104a5e07c vacuum + 1664
7   postgres                            0x0000000104c00670 autovacuum_do_vac_analyze + 136
8   postgres                            0x0000000104bff390 do_autovacuum + 3356
9   postgres                            0x0000000104bfc938 AutoVacWorkerMain + 1244
10  postgres                            0x0000000104bfc444 StartAutoVacWorker + 204
11  postgres                            0x0000000104c1c108 StartAutovacuumWorker + 260
12  postgres                            0x0000000104c1535c sigusr1_handler + 1064
13  libsystem_platform.dylib            0x000000019ae0dc44 _sigtramp + 56
14  postgres                            0x0000000104c16c6c ServerLoop + 364
15  postgres                            0x0000000104c1446c PostmasterMain + 5804
16  postgres                            0x0000000104aff8c8 main + 856
17  libdyld.dylib                       0x000000019ade1f34 start + 4
2021-04-30 10:46:18.903 EDT [22094] LOG:  server process (PID 23447) was terminated by signal 6: Abort trap: 6
2021-04-30 10:46:18.903 EDT [22094] DETAIL:  Failed process was running: autovacuum: VACUUM ANALYZE public.activity_supp_map
2021-04-30 10:46:18.903 EDT [22094] LOG:  terminating any other active server processes
2021-04-30 10:46:18.909 EDT [23449] LOG:  PID 23332 in cancel request did not match any process
2021-04-30 10:46:18.911 EDT [23450] LOG:  PID 23334 in cancel request did not match any process
2021-04-30 10:46:18.912 EDT [22094] LOG:  all server processes terminated; reinitializing
2021-04-30 10:46:18.913 EDT [23451] LOG:  database system was interrupted; last known up at 2021-04-30 10:46:16 EDT
2021-04-30 10:46:18.943 EDT [23451] LOG:  database system was not properly shut down; automatic recovery in progress
2021-04-30 10:46:18.945 EDT [23451] LOG:  redo starts at 5/E9003BB8
2021-04-30 10:46:19.420 EDT [23451] LOG:  invalid record length at 6/3525A78: wanted 24, got 0
2021-04-30 10:46:19.420 EDT [23451] LOG:  redo done at 6/3523A28 system usage: CPU: user: 0.42 s, system: 0.05 s, elapsed: 0.47 s
2021-04-30 10:46:19.423 EDT [22094] LOG:  database system is ready to accept connections
```

```
% uname -a
Darwin Erics-Mac-mini.local 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:06:51 PST 2021; root:xnu-7195.81.3~1/RELEASE_ARM64_T8101 arm64
```

**Comment by hlinnaka** [comment]
Thanks for the testing! Yeah, thing are moving rapidly, I tested this yesterday, and bumped into a bunch of issues. One was fixed in 29f122009a5541017aad520937d1530b79dede14, another was fixed by eeec1a3dcb0b23dbf46e5e6da1af851244be99c9, I think. I just tried again, and this time pg_restore ran to completion. Yay! :-). However, it took many hours; we've spent zero effort on performance so far. On the contrary, we even dial down shared_buffers to 1 MB, to make the buffer cache less efficient, so that the page server gets exercised more.

I didn't test branching while restore is running. I'm sure there are race conditions in that code, we haven't looked into that at all yet, either. Yeah, this is a good test case for that.
